src/pathways/services/pathwayUtils.service.ts

```diff
@@ -0,0 +1,151 @@
+import * as _ from 'lodash';
+import * as fp from 'lodash/fp';
+import * as jsnx from 'jsnetworkx';
+
+export class PathwayUtilsService {
+    // Move to Escher once it's functional
+    DEFAULT_COFACTORS = new Set([
+        'MNXM3',  // 'ATP'
+        'MNXM7',  // 'ADP'
+        'MNXM8',  // 'NAD(+)'
+        'MNXM10',  // 'NADH(2-)'
+        'MNXM5',  // 'NADP(+)
+        'MNXM6',  // 'NADPH'
+        'MNXM51',  // 'GTP'
+        'MNXM30',  // 'GDP'
+        'MNXM12',  // 'CoA'
+        'MNXM80',  // 'UMP(2-)'
+        'MNXM1',  // 'H(+)'
+        'MNXM4',  // 'O2'
+        'MNXM13',  // 'CO(2)'
+        'MNXM2',  // 'H2O'
+    ]);
+
+    // Utility function since the Set interface is rather poor at the moment
+    public setUnion(...sets) {
+        return sets.reduce((ret, nextSet) => {
+            nextSet.forEach((e) => {
+                ret.add(e);
+            });
+            return ret;
+        }, new Set());
+    }
+    /**
+     * Extracts the metabolite names from a list of reactions
+     * The returned list is sorted and unique.
+     * @param reactions 
+     */
+
+    public reactionsMetabolites(reactions) {
+        return _(reactions)
+            .map(x => _.keys(x.metabolites))
+            .flatten()
+            .sortBy()
+            .sortedUniq()
+            .value();
+    }
+
+    /**
+     * Groups reactions by metabolite name
+     * @param reactions 
+     * @param metabolites 
+     */
+    public reactionsByMetabolite(reactions, metabolites) {
+        return _.zipObject(metabolites, _.map(metabolites, (m) => {
+            return _.filter(reactions, (r)=> _.has(r.metabolites, m));
+        }));
+    }
+
+    /**
+     * Returns metabolites with 2 or less reactions
+     * @param reactionsByMetabolite 
+     */
+    public obtainCofactors(reactionsByMetabolite) {
+        return _(reactionsByMetabolite)
+            .pickBy(reacions => reacions.length > 2)
+            .keys()
+            .value();
```

Comment:
Again this seems like `Object.entries(reactionsByMetabolite).filter(([metaboliteName, reactions]) => reactions.length > 2).map(([metaboliteName, reactions]) => metaboliteName)` but I am not sure.